CredentialProvider.Microsoft/CredentialProviders/Vsts/MSAL/MsalTokenProvider.cs

```diff
@@ -79,29 +82,55 @@ public async Task<IMsalToken> AcquireTokenWithDeviceFlowAsync(Func<DeviceCodeRes
 
         public async Task<IMsalToken> AcquireTokenSilentlyAsync(CancellationToken cancellationToken)
         {
-            var publicClient = await GetPCAAsync().ConfigureAwait(false);
-            var accounts = await publicClient.GetAccountsAsync();
+            // first try with the broker
+            IMsalToken token = await AcquireTokenSilentlyInnerAsync(withBroker: true, cancellationToken);
+            if (token != null)
+            {
+                return token;
+            }
 
+            return await AcquireTokenSilentlyInnerAsync(withBroker: false, cancellationToken);
+        }
+
+        private async Task<IMsalToken> AcquireTokenSilentlyInnerAsync(bool withBroker, CancellationToken cancellationToken)
+        {
+            IPublicClientApplication publicClient = null;
             try
             {
+                publicClient = await GetPCAAsync(withBroker, useLocalHost: false).ConfigureAwait(false);
+                var accounts = await publicClient.GetAccountsAsync();
+
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    accounts = accounts.Concat(new [] { PublicClientApplication.OperatingSystemAccount});
+                }
+
```

Comment:
Is there an option for the user to switch account? Beccause what will happen today if the broker is enabled is: 

- user logs in with their Windows account, let's say joe@microsoft.com
- GetAccounts now returns an account for "joe@microsoft.com"
- silent flow always logs in "joe@microsoft.com"


- next time

Reply:
I think @kyle-rader has this flow better structured in the CLI project and can spot some problems. 
This also shows how to use this concept : https://github.com/bgavrilMS/MSAL-SingleAccount-SSO/blob/master/SSO/Program.cs

Reply:
In https://aka.ms/azureauth we currently assume the tool is in use by "corp" accounts, not MSAs, and so for the broker auth mode we will try to use the default OS account if one isn't found first from the cache. 
https://github.com/AzureAD/microsoft-authentication-cli/blob/e3faf814d21441c4353461b23c001382f6dd79f8/src/MSALWrapper/AuthFlow/Broker.cs#L68

Reply:
To Support MSA, this likely needs to change though.

Reply:
@jmyersmsft thoughts on how to approach this?

Reply:
I honestly have no idea. Everything about authentication is always as painful as possible IME

Reply:
In AzureAuth we have a slightly more narrow use case where we expect your machine (if windows) is using a corp-conected account/machine (either AD or AAD), and so if no cached account is found, we automatically try using the Default OS account. We also provide a suffix (domain) filter, to filter cached accounts, and if there's only 1 account after filtering, we will use that - all to avoid account pickers. This is in https://github.com/AzureAD/microsoft-authentication-cli/blob/2836f99cfd1eeb62814056b322765bdcf85a6e86/src/MSALWrapper/PCAWrapper.cs#L117

Reply:
I wonder if I can compare the cached account tenant to the authority...

Reply:
That worked well!